File: House_robber_dp_1d.py
# Dynamic Programing approach
# Time complexity is O(n)
# We will improve upon the space complexity of the previous DP approach. We will reduce the size of our dp array from 2D to 1D
# At any given point we just need the values at the previous step so we dont need to maintain all the values before that.
# This will reduce the Space complexity to O(1) i.e constant

class Solution:
    def rob(self, nums: List[int]) -> int:
        pick,skip = 0,nums[0]
        for i in range(1,len(nums)):
            temp_arr_0 = pick
            pick = max(pick,skip)
            skip = temp_arr_0 + nums[i]
        return max(pick, skip)

File: HouseRobberDP1D.java
//TC = O(n)
//SC = O(n)
class Solution {
    public int rob(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        if(nums.length < 2) return nums[0];
        int n = nums.length;
        int[] dp = new int[n];
        dp[0] = nums[0];
        dp [1] = Math.max(0 + nums[1], dp[0]);
        for(int i = 2; i < n; i++)
        {
            dp[i] = Math.max(dp[i-1], nums[i] + dp[i - 2]);
        }

        return dp[n-1];
    }
}